crates/aptos-rosetta/src/block.rs

```diff
@@ -166,30 +172,73 @@ impl BlockCache {
     }
 
     /// Retrieve the block info for the index
-    ///
-    /// TODO: Improve parallelism and performance
     pub async fn get_block_info(&self, block_index: u64) -> ApiResult<BlockInfo> {
         // If we already have the block info, let's roll with it
-        let (closest_known_block, closest_block_info): (u64, BlockInfo) = {
+        let (start_version, range, traverse_right) = {
             let map = self.blocks.read().unwrap();
             if let Some(block_info) = map.get(&block_index) {
                 return Ok(*block_info);
             }
 
-            // There will always be an index less than the index, since it starts with 0 set
-            let (index, block_info) = map.iter().rev().find(|(i, _)| **i < block_index).unwrap();
-            (*index, *block_info)
+            // Find the closest, left or right, and build up blocks from there
+            let search_left = map.iter().rev().find_map(|(i, info)| {
+                if *i < block_index {
+                    Some((i, info, block_index - i))
+                } else {
+                    None
+                }
+            });
+
+            let search_right = map.iter().find_map(|(i, info)| {
+                if *i > block_index {
+                    Some((i, info, block_index - i))
+                } else {
+                    None
+                }
+            });
+
+            match (search_left, search_right) {
+                (Some((i, info, _)), None) => {
+                    let start = info.end_version + 1;
+                    let range = (i + 1)..=block_index;
+                    (start, range, true)
+                }
+                (None, Some((i, info, _))) => {
+                    let start = info.start_version - 1;
+                    let range = (i - 1)..=block_index;
+                    (start, range, false)
+                }
+                (
+                    Some((left_index, left_info, left_distance)),
+                    Some((right_index, right_info, right_distance)),
+                ) => {
```

Comment:
how is this case even possible?

Reply:
You're searching the cache, and if there's nothing in the cache.  But I seed the cache when I first create it.